### ISSUE TITLE
nth-element selector

### ISSUE THREAD

**Issue by pr8x** [AvaloniaUI/Avalonia#6262]
**Is your feature request related to a problem? Please describe.**

We had some talk about this internally, but I am repeating it here for visibility: It would be nice to have some CSS-like `nth-child` selector that allows you to style every nth descendant.

**Describe the solution you'd like**

`<Style Selector="ListBox ListBoxItem:nth-child(2n+2)" />`

(Would select every 2nd ListBoxItem)

The syntax is debatable, but it's how CSS does it: https://css-tricks.com/how-nth-child-works/

**Describe alternatives you've considered**

WPF's [ItemsControl.AlternationIndex](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.itemscontrol.alternationindex?view=net-5.0) is not flexible enough IMO.

**Additional context**

CC: @grokys

**Comment by maxkatz6** [comment]
Good you mentioned it here. 
I did created simplifed classes setter that can be enabled with attached properties. Will be useful for anybody, who needs it, but can't wait until Avalonia will get proper support for it. 
https://gist.github.com/maxkatz6/124dc79665462982c674b19d788ab6de

**Comment by robloo** [comment]
Could this be used for:
 - Alternative row background/styling in a ListBox (seems yes)
 - Also for rounding corners as discussed here: https://github.com/AvaloniaUI/Avalonia/discussions/5594#discussioncomment-423420. The assumption of course is the buttons are themselves in a ListBox or ItemsRepeater.
     ![](https://user-images.githubusercontent.com/34009058/109704690-e55dae00-7b64-11eb-9bad-84ec06970976.png)

**Comment by wieslawsoltes** [comment]
Based on @maxkatz6 code I have created reusable behavior:

https://github.com/wieslawsoltes/AvaloniaBehaviors/blob/4aa7ce1334b6208ca7321c06783ea5f47ca07d48/src/Avalonia.Xaml.Interactions/Custom/NthChildBehavior.cs#L18

Sample usage:

https://github.com/wieslawsoltes/AvaloniaBehaviors/blob/4aa7ce1334b6208ca7321c06783ea5f47ca07d48/samples/BehaviorsTestApplication/Views/Pages/NthChildBehaviorView.axaml#L17

![image](https://user-images.githubusercontent.com/2297442/125858684-7cc66b59-2b08-4a46-a864-3afc5033f2ef.png)

**Comment by robloo** [comment]
@wieslawsoltes Well, that's extremely useful!

**Comment by wieslawsoltes** [comment]
> @wieslawsoltes @maxkatz6 Well, that's extremely useful!

I've pushed code into NuGet package Avalonia.Xaml.Behaviors 0.10.6.10

**Comment by jmacato** [comment]
@wieslawsoltes i think this would be great if we can integrate this in avalonia somehow :)

**Comment by robloo** [comment]
The only thing to potentially add is some sort of ability to set a modulus. Similar to how WPF handles ItemsControl.AlternationCount and ItemsControl.AlternationIndex.

**Comment by wieslawsoltes** [comment]
> @wieslawsoltes i think this would be great if we can integrate this in avalonia somehow :)

I agree that is really useful feature to have, but we can do it in the meantime using behaviors or by using attached properties.

**Comment by maxkatz6** [comment]
@robloo CSS syntax should be ported to Avalonia. It shouldn't be a big issue to update selectors parser.
Internally we were thinking how we actually should implement index/count for the control items.
In my gist I added attached property, that subscribes on item container generator changes and updates container classes. Enabling this feature for everything would be an overkill. Asking users to enable it or attach behavior property also isn't perfect for something build-in.

Final thoughts were to create something like
```
IChildIndexProvider
{
    (index, totalCount?) GetChildIndex(IControl)
}
```

Which should be implemented on ItemsControl + ItemsPresenter, using information from container generator (which will enable virtualization).
Also, should be implemented on ItemsRepeater with similar behavior.
Questionable, if Panel controls should implement it as well, providing information from flat Children array. The problem with it is "order" - it's not defined for some control types (Grid, Canvas...). But we can implement this interface only on StackPanel and UniformGrid.

Having this interface Selector activator will need to read control's direct parent. If it implements IChildIndexProvider, then we can apply selector. If not - this control will be ignored for this selector.
I think we don't need any kind of "index changed" event. I expect that Any kind of child moves (including virtualization) should invoke AttachedToVisualTree event and it should recalculate styles anyway. But I haven't checked it.

**Comment by pr8x** [comment]
Couldn't it determine index from position in logical tree? 

So we could support something like:

```
<Panel>
  <Button />
  <Button />
</Panel>
```

Are `ListBoxItem` (etc.) actually added to logical tree?

**Comment by maxkatz6** [comment]
@pr8x for panels we can, yes. For lists - no, if they have enabled virtualization.

**Comment by robloo** [comment]
@maxkatz6 Overall I think that's a great design. Only implementing the interface on controls where it makes sense is just fine. If you want to implement this type of functionality using CSS ideas instead of the existing WPF ideas specific to ItemsControl, I also have no concerns with it. It is very much styling related so the CSS methodology seems to make more sense anyway.

I would, however, change the name of things and interface to something closer to `IList` and remove direct mention of child. Keeping things more generic may be useful for future expansion or use in other areas.

Possible interface names IItemList, IElementList, IControlList, etc.

```
IItemList
{
    int IndexOf(IControl control);
    int Count;
}
```
or
```
IControlList
{
    int IndexOfControl(IControl control);
    int ControlCount;
}
```

**Comment by maxkatz6** [comment]
I don't see big problem with index starting with 1

**Comment by RomanSoloweow** [comment]
It would be nice to have at least a child selector
